### PR TITLE
feat(restore): add support for namespace aware restore

### PR DESCRIPTION
**Pull request by shivaji-dgraph**
This commit adds support for namespace aware restore. After this change, we can restore exactly one namespace from a backup of multiple namespaces into a dgraph cluster.

### CODE REVIEW

File: dgraphtest/cluster.go [hunk 1]
```diff
@@ -400,8 +400,7 @@ func (hc *HTTPClient) WaitForTask(taskId string) error {
 
 // Restore performs restore on Dgraph cluster from the given path to backup
 func (hc *HTTPClient) Restore(c Cluster, backupPath string,
-	backupId string, incrFrom, backupNum int, encKey string) error {
-
+	backupId string, incrFrom, backupNum int, encKey string, namespace uint64) error {
```
Comment (mangalaman93) [inline dgraphtest/cluster.go:403]: let's call it `fromNamespace` instead, same everywhere below

File: graphql/admin/endpoints_ee.go [hunk 2]
```diff
@@ -143,6 +143,11 @@ const adminTypes = `
 		Set to true to allow backing up to S3 or Minio bucket that requires no credentials.
 		"""
 		anonymous: Boolean
+
+		"""
```
Comment (mangalaman93) [inline graphql/admin/endpoints_ee.go:147]: this needs more explanation

File: systest/online-restore/namespace-aware/restore_test.go [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+
+	ns1, err := hc.AddNamespace()
+	require.NoError(t, err)
+
+	require.NoError(t, gc.LoginIntoNamespace(context.Background(), "groot", "password", ns1))
+	require.NoError(t, gc.SetupSchema(emailSchema))
+	rdf := ""
+	for i := 1; i <= 100; i++ {
+		rdf = rdf + fmt.Sprintf("_:a%v <email> \"example%v@gmail.com\" .	\n", i, i)
+
```
Comment (mangalaman93) [inline systest/online-restore/namespace-aware/restore_test.go:65]: nit: remove the newline

File: systest/online-restore/namespace-aware/restore_test.go [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+	require.NoError(t, hc.LoginIntoNamespace("groot", "password", 0))
+	require.NoError(t, hc.Backup(c, false, dgraphtest.DefaultBackupDir))
+
+	require.NoError(t, hc.Restore(c, dgraphtest.DefaultBackupDir, "", 0, 0, dgraphtest.EncKeyMountPath, ns1))
+	require.NoError(t, dgraphtest.WaitForRestore(c))
+
+	gc, cleanup, err = c.Client()
+	require.NoError(t, err)
+	defer cleanup()
+
```
Comment (mangalaman93) [inline systest/online-restore/namespace-aware/restore_test.go:104]: nit: remove the newline

File: systest/online-restore/namespace-aware/restore_test.go [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+	require.NoError(t, dgraphtest.CompareJSON(`{"allMails":[{"count":100}]}`, string(resp.Json)))
+
+	query = `{
+		allMails(func: has(name)) {
+			   count(uid)
+			 }
+	 }`
+	resp, err = gc.Query(query)
+	require.NoError(t, err)
+	require.NoError(t, dgraphtest.CompareJSON(`{"allMails":[{"count":0}]}`, string(resp.Json)))
```
Comment (mangalaman93) [inline systest/online-restore/namespace-aware/restore_test.go:124]: can we also have emails in ns2 in the test?

File: systest/online-restore/namespace-aware/restore_test.go [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+			   count(uid)
+			 }
+	 }`
+	resp, err := gc.Query(query)
+	require.NoError(t, err)
+	// there should be 100 emails in DB.
+	require.NoError(t, dgraphtest.CompareJSON(`{"allMails":[{"count":100}]}`, string(resp.Json)))
+
+	query = `{
+		allMails(func: has(name)) {
```
Comment (mangalaman93) [inline systest/online-restore/namespace-aware/restore_test.go:118]: change this to allNames and you could say `const query = ""`

File: systest/online-restore/namespace-aware/restore_test.go [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+			   count(uid)
+			 }
+	 }`
+	resp, err = gc.Query(query)
+	require.NoError(t, err)
+	require.NoError(t, dgraphtest.CompareJSON(`{"allMails":[{"count":0}]}`, string(resp.Json)))
+}
+
+func TestNamespaceAwareRestoreOnMultipleGroup(t *testing.T) {
```
Comment (mangalaman93) [inline systest/online-restore/namespace-aware/restore_test.go:127]: nit: MultipleGroup*s*

File: systest/online-restore/namespace-aware/restore_test.go [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+	conf := dgraphtest.NewClusterConfig().WithNumAlphas(6).WithNumZeros(6).
+		WithReplicas(3).WithACL(20 * time.Hour).WithEncryption().WithUidLease(10000)
+	c, err := dgraphtest.NewLocalCluster(conf)
+	x.Panic(err)
+	if err := c.Start(); err != nil {
+		c.Cleanup(true)
```
Comment (mangalaman93) [inline systest/online-restore/namespace-aware/restore_test.go:133]: fix this too

File: systest/online-restore/namespace-aware/restore_test.go [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+	})
+	require.NoError(t, err)
+
+	fmt.Println("Inserted email predicate into namespace 0")
+
+	require.NoError(t, gc.Login(context.Background(), "groot", "password"))
+	require.NoError(t, hc.LoginIntoNamespace("groot", "password", 0))
+
+	// creating multiple namespaces and adding email predicate into them
+	// so email predicte distributed into both groups
```
Comment (mangalaman93) [inline systest/online-restore/namespace-aware/restore_test.go:201]: can you write down in comments what is the distribution of predicates in the groups at this point?

File: systest/online-restore/namespace-aware/restore_test.go [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+				CommitNow: true,
+			})
+		}
+		require.NoError(t, hc.LoginIntoNamespace("groot", "password", 0))
+	}
+
+	require.NoError(t, hc.LoginIntoNamespace("groot", "password", 0))
+	require.NoError(t, hc.Backup(c, false, dgraphtest.DefaultBackupDir))
+
+	fmt.Println("Restoring only namespace ", ns1)
```
Comment (mangalaman93) [inline systest/online-restore/namespace-aware/restore_test.go:223]: t.Logf please

File: systest/online-restore/namespace-aware/restore_test.go (hunk 7 above)
Comment (mangalaman93) [inline systest/online-restore/namespace-aware/restore_test.go:127]: we should also add a test with incremental restore as well

File: worker/online_restore.go [hunk 11]
```diff
@@ -309,6 +309,11 @@ func handleRestoreProposal(ctx context.Context, req *pb.RestoreRequest, pidx uin
 		return errors.Errorf("backup manifest does not contain information for group ID %d", req.GroupId)
 	}
 
+	// Make sure to assign the correct alpha group when restoring predicates
```
Comment (mangalaman93) [inline worker/online_restore.go:312]: mention "namespace aware restore" here

File: worker/online_restore.go [hunk 12]
```diff
@@ -309,6 +309,11 @@ func handleRestoreProposal(ctx context.Context, req *pb.RestoreRequest, pidx uin
...
+	// from any namespace to namespace 0.
+	if req.Namespace != -1 {
```
Comment (mangalaman93) [inline worker/online_restore.go:314]: check for < 0

File: worker/online_restore.go [hunk 13]
```diff
@@ -309,6 +309,11 @@ func handleRestoreProposal(ctx context.Context, req *pb.RestoreRequest, pidx uin
...
+		restorePreds = addPredIntoRestorePreds(restorePreds, req.Namespace)
```
Comment (mangalaman93) [inline worker/online_restore.go:315]: we should also add test for namespace 0 -> namespace 0 restore

File: worker/online_restore.go [hunk 14]
```diff
@@ -594,3 +599,15 @@ func RunOfflineRestore(dir, location, backupId, keyFile string, key x.Sensitive,
 	// TODO: Fix this return value.
 	return LoadResult{Version: manifest.ValidReadTs()}
 }
+
+func addPredIntoRestorePreds(restorePreds []string, namespace int64) []string {
```
Comment (mangalaman93) [inline worker/online_restore.go:603]: we should add unit tests for this function

Comment (mangalaman93) [reply worker/online_restore.go:603]: namespace -> fromNamespace and make the type same as `ns`

File: worker/online_restore.go [hunk 15]
```diff
@@ -594,3 +599,15 @@ func RunOfflineRestore(dir, location, backupId, keyFile string, key x.Sensitive,
...
+	var newRestorePreds []string
```
Comment (mangalaman93) [inline worker/online_restore.go:604]: you could the the trick for filter in slice

## Filtering without allocating

This trick uses the fact that a slice shares the same backing array and capacity as the original, so the storage is reused for the filtered slice. Of course, the original contents are modified.

b := a[:0]
for _, x := range a {
	if f(x) {
		b = append(b, x)
	}
}

File: worker/restore_map.go [hunk 16]
```diff
@@ -323,6 +325,13 @@ func (m *mapper) processReqCh(ctx context.Context) error {
 			return errors.Wrap(err, "fromBackupKey")
 		}
 
+		// Verifying whether the user has requested a namespace-aware restore
+		isNameSpaceAwareRestore := in.namespace > 0
```
Comment (mangalaman93) [inline worker/restore_map.go:329]: what if it is equal to zero

File: worker/restore_map.go [hunk 17]
```diff
@@ -333,7 +342,7 @@ func (m *mapper) processReqCh(ctx context.Context) error {
 
 		// Update the local max uid and max namespace values.
 		maxUid = x.Max(maxUid, parsedKey.Uid)
-		maxNs = x.Max(maxNs, ns)
+		maxNs = x.Max(maxNs, 0)
```
Comment (mangalaman93) [inline worker/restore_map.go:345]: this doesn't look right for the normal case

File: worker/restore_map.go [hunk 18]
```diff
@@ -342,11 +351,24 @@ func (m *mapper) processReqCh(ctx context.Context) error {
 			return nil
 		}
 
+		if isNameSpaceAwareRestore {
+			// update a key to default namespace if user has requested namespace aware restore
+			// i.e 1-key => 0-key
+			parsedKey.Attr = strings.Replace(parsedKey.Attr, strconv.FormatUint(ns, 16), "0", 1)
```
Comment (mangalaman93) [inline worker/restore_map.go:357]: we should not use strings.Replace. Predicates could have a the namespace number too. See if we could add a test with such a predicate.

**Review by mangalaman93 — CHANGES_REQUESTED** [review]
I don't see any code for moveTablet. How is this working?

File: dgraphtest/cluster.go [hunk 19]
```diff
@@ -604,18 +613,46 @@ func (hc *HTTPClient) GetZeroState() (*LicenseResponse, error) {
...
 	var stateResponse LicenseResponse
 	if err := json.Unmarshal(body, &stateResponse); err != nil {
-		return nil, errors.New("error unmarshaling zero state response")
+		return nil, errors.Wrapf(err, "error unmarshaling zero state response")
 	}
 
 	return &stateResponse, nil
 }
 
+func (hc *HTTPClient) GetState() (*pb.MembershipState, error) {
```
Comment (mangalaman93) [inline dgraphtest/cluster.go:634]: we probably don't need this any more

File: worker/online_restore.go [hunk 20]
```diff
@@ -594,3 +599,16 @@ func RunOfflineRestore(dir, location, backupId, keyFile string, key x.Sensitive,
 	// TODO: Fix this return value.
 	return LoadResult{Version: manifest.ValidReadTs()}
 }
+
+// addPredIntoRestorePreds
```
Comment (mangalaman93) [inline worker/online_restore.go:603]: explain what this function is doing and why.

File: worker/restore_map.go [hunk 21]
```diff
@@ -323,6 +326,10 @@ func (m *mapper) processReqCh(ctx context.Context) error {
 			return errors.Wrap(err, "fromBackupKey")
 		}
 
+		if in.oneNamespaceOnly && ns != in.fromNamespace {
```
Comment (mangalaman93) [inline worker/restore_map.go:329]: add a comment

File: worker/restore_map.go [hunk 22]
```diff
@@ -333,7 +340,11 @@ func (m *mapper) processReqCh(ctx context.Context) error {
 
 		// Update the local max uid and max namespace values.
 		maxUid = x.Max(maxUid, parsedKey.Uid)
-		maxNs = x.Max(maxNs, ns)
+		if in.oneNamespaceOnly {
```
Comment (mangalaman93) [inline worker/restore_map.go:343]: add a comment here, and I think this is not right. Try performing a namespace aware restore and then create an existing namesapce. It should not work.

File: worker/restore_map.go [hunk 23]
```diff
@@ -468,6 +493,48 @@ func (m *mapper) processReqCh(ctx context.Context) error {
 			version := kv.Version
 			kv.Version = m.restoreTs
 			kv.Key = restoreKey
+			changeNameSpaceToDefault := func(parsedKey x.ParsedKey) error {
```
Comment (mangalaman93) [inline worker/restore_map.go:496]: changeNameSpaceToDefault is not self explanatory

**Review by mangalaman93 — CHANGES_REQUESTED** [review]
More todos:
1. I do not like the API, oneNamespaceOnly flag. we need to think more.
2. Uprgade tests will fail because of the change in cluster.go restore function. It should support both old and new API
3. We should test against a fresh cluster when zero doesn't have any tablets

File: graphql/admin/endpoints_ee.go [hunk 24]
```diff
@@ -143,6 +143,19 @@ const adminTypes = `
...
+		oneNamespaceOnly should be set if only one namespace is to be restored from
+		the backup. The namespace value should be set into the fromNamespace parameter.
+		The given namespace is always restored into namespace 0 in the cluster.
+		"""
+		oneNamespaceOnly: Boolean
```
Comment (sanjayk-github-dev) [inline graphql/admin/endpoints_ee.go:152]: Do we need `oneNamespaceOnly`? Can we use `fromNamespace > 0` to imply the same?

File: dgraphtest/cluster.go [hunk 25]
```diff
@@ -386,24 +388,41 @@ func (hc *HTTPClient) Restore(c Cluster, backupPath string,
 		return errors.New("incremental restore is not supported by the cluster")
 	}
 
-	var varPart, queryPart string
+	isNamespaceAwareRestorSupported, err := IsHigherVersion(c.GetVersion(), "ba727644664138d3ddba96a21348612b2bec2d88")
```
Comment (mangalaman93) [inline dgraphtest/cluster.go:391]: isNamespaceAwareRestor*e*Supported

File: dgraphtest/cluster.go [hunk 26]
```diff
@@ -386,24 +388,41 @@ func (hc *HTTPClient) Restore(c Cluster, backupPath string,
...
+			encryptionKeyFile: $encKey, }) {
 			code
 			message
 		}
-	}`, varPart, queryPart)
-	vars := map[string]interface{}{"location": backupPath, "backupId": backupId,
-		"backupNum": backupNum, "encKey": encKey}
+	}`, varPart.String(), queryPart.String())
+	vars := map[string]interface{}{"location": backupPath, "backupId": backupId, "backupNum": backupNum,
+		"encKey": encKey, "oneNamespaceOnly": oneNamespaceOnly, "fromNamespace": fromNamespace}
```
Comment (mangalaman93) [inline dgraphtest/cluster.go:418]: we don't need to have the keys here "oneNamespaceOnly" & "fromNamespace"

**Review by mangalaman93 — CHANGES_REQUESTED** [review]
We should consider having a completely new API for this.

File: systest/online-restore/namespace-aware/restore_test.go [hunk 27]
```diff
@@ -0,0 +1,219 @@
...
+			require.NoError(t, dgraphtest.WaitForRestore(c))
+
+			gc, cleanup, err = c.Client()
+			require.NoError(t, err)
+			defer cleanup()
+
+			// Only the namespace '0' should have data
+			require.NoError(t, gc.LoginIntoNamespace(context.Background(),
+				dgraphtest.DefaultUser, dgraphtest.DefaultPassword, x.GalaxyNamespace))
+			const query = `{
```
Comment (mangalaman93) [inline systest/online-restore/namespace-aware/restore_test.go:86]: nit: format query

**Review by mangalaman93 — APPROVED** [review]

**Review by harshil-goel — APPROVED** [review]

**Pull request merged**